### PR TITLE
doc/developer: add "Catalog tables for Optimizer Notices" design doc

### PR DESCRIPTION
**Pull request by aalexandrov**
Design doc for the infrastructure requested in #21513.

[Rendered version](https://github.com/aalexandrov/materialize/blob/issue_21513_design/doc/developer/design/20231113_optimizer_notice_catalog.md).

### Motivation

  * This PR adds a known-desirable feature.

### Tips for reviewers

Still needs some sections to be filled in. I hope to move it out from draft mode later today.

- [x] @ggnall / @sthm please make sure that the problem statement and the UX diagram are aligned with your vision / feature requests.
- [x] @ggnall I'm also looking into guidance about the **RBAC issues** alternatives from the **Open questions** section.
- [x] @ggevay as the primary reviewer of #23360 the most relevant parts for you are the **Catalog Schema** and **Schema Maintenance** parts of the **Solution Proposal**.
- [ ] @ParkMyCar mainly I want to double-check for possible performance regressions in the way I intend to do **`mz_all_optimizer_notices` maintenance**.
- [x] @aljoscha please review the **Changes required for Platform V2** and let me know if this aligns with your vision for distributed `cooordd`.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - There are no user-facing behavior changes.

### CODE REVIEW

File: doc/developer/design/20231113_optimizer_notice_catalog.md
Comment (ggevay) [inline doc/developer/design/20231113_optimizer_notice_catalog.md:151]: "Schema Maintenance" sounds to me as if it were about changes to the schema itself, such as adding new columns.

File: doc/developer/design/20231113_optimizer_notice_catalog.md
Comment (ggevay) [inline doc/developer/design/20231113_optimizer_notice_catalog.md:173]: during the next restart

File: doc/developer/design/20231113_optimizer_notice_catalog.md
Comment (ggevay) [inline doc/developer/design/20231113_optimizer_notice_catalog.md:176]: it in the

File: doc/developer/design/20231113_optimizer_notice_catalog.md
Comment (ggevay) [inline doc/developer/design/20231113_optimizer_notice_catalog.md:186]: (For the corrective action of IndexTooWideForLiteralConstraints, we should mention that the user might want to drop the old index _if it is not used by some other queries_.)

File: doc/developer/design/20231113_optimizer_notice_catalog.md
Comment (ggevay) [inline doc/developer/design/20231113_optimizer_notice_catalog.md:202]: I've been thinking that maybe instead of "Optimizer notices" we should call these things "Optimization notices", because some of them are actually emitted not by the optimizer. For example, `DroppedInUseIndex` is created in `sequence_drop_common`.

File: doc/developer/design/20231113_optimizer_notice_catalog.md
Comment (ggevay) [inline doc/developer/design/20231113_optimizer_notice_catalog.md:216]: We are recording `action_type` because at some point in the future we might want to add a button that simply executes it if it is `sql_statements`?

File: doc/developer/design/20231113_optimizer_notice_catalog.md
Comment (ggevay) [inline doc/developer/design/20231113_optimizer_notice_catalog.md:218]: I'm thinking about the exact connection between `object_id` and `dependency_ids`. For an example, let's consider the `CommonArrangedInput` notice (I just made up this name; it's https://github.com/MaterializeInc/materialize/issues/20802).

In this example, would `dependency_ids` list all those objects that arrange the same one input? But then if one of these objects is dropped, the notice might still be valid if originally there were more than two objects arranging the same input. It becomes invalid only if all but one of the ids in `dependency_ids` is dropped.

Another question for this example: what will be the object id? Will we have one row for each object in `dependency_ids`?

**Review by ggevay — COMMENTED** [review]
I wrote some comments. I'm still going through the design doc.

File: doc/developer/design/20231113_optimizer_notice_catalog.md
Comment (aalexandrov) [reply doc/developer/design/20231113_optimizer_notice_catalog.md]: I don't think that we are looking to maintain these notices in the same table. My impression was that the current iteration focuses precisely on the notices emitted by the optimizer as well as other notices that we can add in the future.

For example, dropping an index that is still in use will create dangling `import_ids` in some of the plans of running indexes, so we can have other notices attached to the specific index or materialized view that depends on that index.

File: doc/developer/design/20231113_optimizer_notice_catalog.md
Comment (aalexandrov) [reply doc/developer/design/20231113_optimizer_notice_catalog.md:220]: Yes (this request came from @sthm). A button probably won't cut it until we have zero-downtime `REPLAN` which seems some way down the road. However, I guess even now it is better UX if the customer can copy-paste a valid SQL string the notice into their dbt project or whatever else they use to populate their catalog.

File: doc/developer/design/20231113_optimizer_notice_catalog.md
Comment (aalexandrov) [reply doc/developer/design/20231113_optimizer_notice_catalog.md:187]: Are you with OK with just emitting a notice for indexes that are not used at all? Then handling the `IndexTooWideForLiteralConstraints` and re-creating the objects might result in another notice.

File: doc/developer/design/20231113_optimizer_notice_catalog.md
Comment (aalexandrov) [reply doc/developer/design/20231113_optimizer_notice_catalog.md]: I will rename this to "Optimization notices".

File: doc/developer/design/20231113_optimizer_notice_catalog.md
Comment (aalexandrov) [reply doc/developer/design/20231113_optimizer_notice_catalog.md:222]: Maybe another way to model this is to have `dependant_ids` (as a separate relation or a list) instead of `object_id`. This will contain one entry for notices scoped to a single element and multiple entries for things like `CommonArrangedInput`.

File: doc/developer/design/20231113_optimizer_notice_catalog.md
Comment (ggevay) [reply doc/developer/design/20231113_optimizer_notice_catalog.md:187]: This is tricky, because we'd like to keep notices quite high-signal, i.e., false positives should be rare. But `IndexNotUsed` would definitely show a false positive for each serving index, if we implement it by only looking at static objects. So, I'd say we should add this notice only when we'll also have the last usage time of each index implemented (https://github.com/MaterializeInc/materialize/issues/20877), because then we can take that also into account when deciding whether to show an `IndexNotUsed`.

Comment (ggevay) [reply doc/developer/design/20231113_optimizer_notice_catalog.md:187]: > Are you with OK with just emitting a dedicated notice for indexes that are not used at all?

Ok! And then when we have `IndexNotUsed` later implemented in the above way, then these things will fall into place.

File: doc/developer/design/20231113_optimizer_notice_catalog.md
Comment (ggevay) [reply doc/developer/design/20231113_optimizer_notice_catalog.md:222]: Hmm, ok, so this will be used by Rust code when computing table updates, and not by some SQL in a view that just simply hides the row when any (or all) of the things in `dependency_ids` is gone, right?

Comment (ggevay) [reply doc/developer/design/20231113_optimizer_notice_catalog.md:222]: > Maybe another way to model this is to have dependant_ids (as a separate relation or a list) instead of object_id. This will contain one entry for notices scoped to a single element and multiple entries for things like CommonArrangedInput.

Sounds good.

Comment (aalexandrov) [reply doc/developer/design/20231113_optimizer_notice_catalog.md:222]: Yes, sadly I don't see a good way around that. The `BuiltinTable` API is written that way - it is meant to be a store for side-effects that reflects certain parts of the in-memory state of the system. The adapter code does not have access to the contents of any `BuiltinTable` - it can only emit a retraction (based on its current state) and create an insertion against table.

@ParkMyCar please disagree with me here if this is not correct, but this was my impression from reading the code.

File: doc/developer/design/20231113_optimizer_notice_catalog.md
Comment (aalexandrov) [reply doc/developer/design/20231113_optimizer_notice_catalog.md]: Renamed this to "Notice Generation and Management".

File: doc/developer/design/20231113_optimizer_notice_catalog.md
Comment (aalexandrov) [inline doc/developer/design/20231113_optimizer_notice_catalog.md:325]: @ParkMyCar I guess since you are currently in the SQL Council I need your approval on this as well.

File: doc/developer/design/20231113_optimizer_notice_catalog.md
Comment (aalexandrov) [reply doc/developer/design/20231113_optimizer_notice_catalog.md:187]: > So, I'd say we should add this notice only when we'll also have the last usage time of each index implemented (https://github.com/MaterializeInc/materialize/issues/20877), because then we can take that also into account when deciding whether to show an IndexNotUsed.

The design also provisions for ignoring a notice in a way that will survive restarts. If this is the case the user can just manually ignore all false positive notices for `IndexNotUsed`.

File: doc/developer/design/20231113_optimizer_notice_catalog.md
Comment (ggnall) [inline doc/developer/design/20231113_optimizer_notice_catalog.md:18]: This section looks good to me. For our current service model this also means Field Engineering can't look up prior notices and can only rely on asking for (or running) realtime EXPLAINs.

File: doc/developer/design/20231113_optimizer_notice_catalog.md
Comment (ggnall) [inline doc/developer/design/20231113_optimizer_notice_catalog.md:515]: This seems the most straightforward for an FE-targeted first version.

File: doc/developer/design/20231113_optimizer_notice_catalog.md
Comment (ggnall) [inline doc/developer/design/20231113_optimizer_notice_catalog.md:518]: Not following why this must be the case? UX seems like the primary interface, but I'm not sure why SQL wouldn't be supported.

Comment (aalexandrov) [reply doc/developer/design/20231113_optimizer_notice_catalog.md:518]: I didn't consider that regular users where `User::is_external_admin()` is true will also be able to see the notices through the SQL console. I updated the text a bit to reflect that.

**Review by ggevay — COMMENTED** [review]
A few more minor comments

**Review by sthm — APPROVED** [review]
Lgtm, thanks for working on this!

File: doc/developer/design/20231113_optimizer_notice_catalog.md
Comment (ParkMyCar) [reply doc/developer/design/20231113_optimizer_notice_catalog.md:222]: Personally I would love if we could unify the two, specifically if we could also eagerly retract from `mz_all_optimizer_notices` for the `dependency_ids` column.

File: doc/developer/design/20231113_optimizer_notice_catalog.md
Comment (ParkMyCar) [inline doc/developer/design/20231113_optimizer_notice_catalog.md:290]: Can you go into more detail about how we would issue retractions? For example, if a dependent object was dropped, or an object was renamed, how do we issue a retraction, even by using the Catalog? The previous dependent ID is gone from the catalog, as is the previous name.

If there's already code for this in your PR that handles this scenario, I'm more than happy to read through that :)

File: doc/developer/design/20231113_optimizer_notice_catalog.md
Comment (ParkMyCar) [inline doc/developer/design/20231113_optimizer_notice_catalog.md:299]: Do we have an estimate to how many entries might exist in `mz_all_optimizer_notices`? We should double check that filter pushdown works on a column, preferably on the `object_id` column which I'm guessing is what users will mostly filter on

File: doc/developer/design/20231113_optimizer_notice_catalog.md
Comment (ParkMyCar) [inline doc/developer/design/20231113_optimizer_notice_catalog.md:334]: Generally `ALTER` commands take the format `ALTER <type> <name> <field>` (e.g. [alter cluster](https://materialize.com/docs/sql/alter-cluster/) or [alter index](https://materialize.com/docs/sql/alter-index/)) what do you think about the following format?
```
ALTER NOTICE <fingerprint> SET STATUS = '<status>';
ALTER NOTICE <finterprint> RESET STATUS;
```

File: doc/developer/design/20231113_optimizer_notice_catalog.md
Comment (ParkMyCar) [inline doc/developer/design/20231113_optimizer_notice_catalog.md:462]: What about durably storing the notices, or some portion of the notice, in the Stash, instead of a builtin table/source? This is what we do for object comments, and it might alleviate some of the problems around retractions?

**Review by ParkMyCar — COMMENTED** [review]
I love this idea! I need to understand a bit better how retractions from `mz_all_optimizer_notices`, would work. More than happy to chat in a huddle or zoom if that would help!

File: doc/developer/design/20231113_optimizer_notice_catalog.md
Comment (aalexandrov) [reply doc/developer/design/20231113_optimizer_notice_catalog.md:462]: > What about durably storing the notices, or some portion of the notice, in the Stash, instead of a builtin table/source? 

I think the main reason not to do that is that that optimizer notices are a function of the catalog state, more specifically a lot of them would be a function of properties of the plans for materialized views and indexes created during bootstrap.

This means that when we restart / upgrade materialize the plans might change. We would need to somehow discard all optimizer notices created before the restart and create new set of notice entries for the builtin source. Given these semantics, I thought that maintaining the optimizer notices in a builtin table is sufficient for now.

File: doc/developer/design/20231113_optimizer_notice_catalog.md
Comment (aalexandrov) [reply doc/developer/design/20231113_optimizer_notice_catalog.md:290]: There are two ways we remove remove / supress notices from the `mz_optimizer_notices` view:

1. When an entry referenced in `object_id` is deleted, we remove the notice [at the end of `Catalog::transact`](https://github.com/MaterializeInc/materialize/pull/23360/files#diff-6de498e18d68f9aa0a688e2b6bef9b0bc4057ba40284926d94126c93670ec481R1193-R1201).
2. When an entry referenced in `dependency_ids` is deleted, the [current `mz_optimizer_notices` view definition](https://github.com/MaterializeInc/materialize/pull/23360/files#diff-e31a13cbb126bb7fe119e3bf9790c20249c84db6408ea6bf548368c64428e8f6R51-R69) will filter out the enclosing notice.

In the long term I would like to emit eager retractions also for (2), but for the the sake of merging an MVP that would allow me to put together an internal console app and expose the feature to field engineering for beta testing I thought this discrepancy won't hurt too much.

**Comment by aalexandrov** [comment]
> I need to understand a bit better how retractions from `mz_all_optimizer_notices`, would work. More than happy to chat in a huddle or zoom if that would help!

I have an MVP open in MaterializeInc/materialize#23360. I think it's best to have a huddle and go over the main parts of the PR together because I plan to change some things before I merge this PR and align the mechanisms used to retract invalid notices shortly after that with a follow-up PR.

**Review by ParkMyCar — APPROVED** [review]
@aalexandrov and I chatted over Zoom and we're aligned on this design!

The follow ups we have for later versions of the feature are:
* Eagerly retracing notices when a dependency goes away, e.g. an index whose key was too wide gets dropped. We think this can be handled by adding a `dependencies` field to [`CatalogPlans`](https://github.com/MaterializeInc/materialize/blob/859414d48fcba31ffb703953724b682dcb4e5623/src/adapter/src/catalog.rs#L157-L161), and then add a check to the Coordinator's consistency checks.
* Handling of the ignored/resolved state. This is a bit trickier because we need to survive restarts of envd, recreation of objects, and ideally renaming of the object, or the schema/database it exists in. There are some ideas on how to "fingerprint" this ignored/resolved state, but punting on it for now as it's not required for V0.

**Comment by aalexandrov** [comment]
@ParkMyCar I updated the design doc. If I can squeeze in the refactor to do the first bit (Eagerly retracing notices when a dependency goes away) I might tag you for a fast follow-up PR tomorrow.

**Pull request merged**
